internal/opcua/client.go

```diff
@@ -0,0 +1,221 @@
+package opcua_client
+
+import (
+	"context"
+	"fmt"
+	"github.com/gopcua/opcua"
+	"github.com/gopcua/opcua/ua"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+type OpcUAWorkarounds struct {
+	AdditionalValidStatusCodes []string `toml:"additional_valid_status_codes"`
+}
+
+type OpcUAClientConfig struct {
+	Endpoint       string
+	SecurityPolicy string
+	SecurityMode   string
+	Certificate    string
+	PrivateKey     string
+	Username       string
+	Password       string
+	AuthMethod     string
+	ConnectTimeout config.Duration
+	RequestTimeout config.Duration
+
+	Workarounds OpcUAWorkarounds
+}
+
+func (o *OpcUAClientConfig) Validate() error {
+	err := o.validateEndpoint()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *OpcUAClientConfig) validateEndpoint() error {
+	if o.Endpoint == "" {
+		return fmt.Errorf("endpoint url is empty")
+	}
+
+	_, err := url.Parse(o.Endpoint)
+	if err != nil {
+		return fmt.Errorf("endpoint url is invalid")
+	}
+
+	//search security policy type
+	switch o.SecurityPolicy {
+	case "None", "Basic128Rsa15", "Basic256", "Basic256Sha256", "auto":
+		// Valid security policy type - do nothing.
+	default:
+		return fmt.Errorf("invalid security type '%s' in '%s'", o.SecurityPolicy, o.Endpoint)
+	}
+	//search security mode type
+	switch o.SecurityMode {
+	case "None", "Sign", "SignAndEncrypt", "auto":
+		// Valid security mode type - do nothing.
+	default:
+		return fmt.Errorf("invalid security type '%s' in '%s'", o.SecurityMode, o.Endpoint)
+	}
+	return nil
+}
+
+// ConnectionState used for constants
+type ConnectionState int
+
+const (
+	//Disconnected constant State 0
+	Disconnected ConnectionState = iota
+	//Connecting constant State 1
+	Connecting
+	//Connected constant State 2
+	Connected
+)
+
+type OpcUAClient struct {
+	Config *OpcUAClientConfig
+	Log    telegraf.Logger
+
+	State  ConnectionState
+	Client *opcua.Client
+
+	opts  []opcua.Option
+	codes []ua.StatusCode
+}
```

Comment:
Client and State need to be exported, because the OpcUAClient is now in the common package and SubscribeClient and ReadClient need access to it.

I do not entirely understand what you mean with generating the Client from a config and not needing to store it internally. We need the client in memory, because we need to maintain a connection to the server.